### PR TITLE
1361 add title and export date to exported chart

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
[1361 - Change to Exported Visuals](https://github.com/beyondessential/tupaia-backlog/issues/1361)

### Changes:

1. Add title at the top and export date at the bottom of exports excel files, whicn includes: 
    - Raw data download 
    - Custom data download 
    - Matrix data exports

2. Refactor the code in `packages/meditrak-server/src/routes/exportSurveyResponses.js`

### CODE REVIEW

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:305]: Add two `[]` for spacing between the table and the export date.

File: packages/web-config-server/src/export/excelFormatters/formatMatrixDataForExcel.js
Comment (Alastair-L) [inline packages/web-config-server/src/export/excelFormatters/formatMatrixDataForExcel.js:92]: Is there any way to consolidate this code with the what's in the meditrak-server? Perhaps an 'buildExcelExport' util or something, which receives the data as an array and some metadata, and adds the title and export date and builds the report? What do you think about that?

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:92]:
```suggestion
    const reportName = viewId && (await models.dashboardReport.findById(viewId)).viewJson.name;
```
?

**Review by Alastair-L — CHANGES_REQUESTED** [review]
Looks good! Just a few suggestions, feel free to take them or leave them. 


Also @nick-beyondessential as our resident export expert would you be able to take a look at https://github.com/beyondessential/tupaia/pull/1627/files#diff-a6df91d7f12e45a71dc35af6198eb68f8ff05d8b0e5fc9f38d41290abe941caaR94-R98?

File: packages/web-config-server/src/export/excelFormatters/formatMatrixDataForExcel.js
Comment (billli0) [reply packages/web-config-server/src/export/excelFormatters/formatMatrixDataForExcel.js:92]: That's a good idea. Will take that.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js:92]: Look nicer! That inspires me to refactor the code.

**Comment by nick-beyondessential** [comment]
> Looks good! Just a few suggestions, feel free to take them or leave them.
> 
> Also @nick-beyondessential as our resident export expert would you be able to take a look at https://github.com/beyondessential/tupaia/pull/1627/files#diff-a6df91d7f12e45a71dc35af6198eb68f8ff05d8b0e5fc9f38d41290abe941caaR94-R98?
@Alastair-L I think this is a fair way to go, the idea behind formattedData is that it is a manipulatable array of arrays and this is the last step before export so shouldn't clash with other transformations.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js:92]: Seems like the name couldn't be fetched from the object in this way unfortunately. But I did some refactoring for the code below. Hope that can add value.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js]:
```suggestion
    const reportName = viewId && (await models.dashboardReport.findById(viewId)).viewJson.name;
```
What about this with the bracket moved? I think you were trying to get `viewJson` of a promise.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:114]: could you refactor this further to:
```
const getParametersFromInput = (variablesExtractor, a, b, ...) => {
 if (a) return variablesExtractor.getVariablesByCountryId(a);
...
else{ throw }
}
```
and replace the logic here with:
```javascript
const { country, entities, ... } = await getParametersFromInput(variablesExtractor, entitityCode, countryId, ...);
```

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:126]: Great refactor!!

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:9]: You'll need to change this I think

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:39]: I reckon these should all have a consistent return type if possible (`{country, countryId, entities}` or just `{ country, entities }`, plus `surveyResponse` and `surveyId` optionally). You could use [typedef](https://jsdoc.app/tags-typedef.html) to ensure this consistency is documented, but it's not essential.

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:30]: I think all the jsdocs in this file could be fixed :).

**Review by Alastair-L — CHANGES_REQUESTED** [review]
Love the refactoring effort! Much better to my eyes. A couple of (optional) further improvements and then just fixing up the jsdocs. 

Also, it's worth mentioning that this is a great candidate for unit tests, but they're also optional from my perspective.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js:114]: Sure! This looks better.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js]: Nice! Works now!

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (billli0) [reply packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js]: I forgot to do so. My bad.

File: packages/utils/src/__tests__/period/convertDateRangeToPeriods.test.js
Comment (Alastair-L) [inline packages/utils/src/__tests__/period/convertDateRangeToPeriods.test.js:96]: Nice use of parameterized tests🎉

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (billli0) [reply packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:29]: Good suggestion! Didn't know any jsdoc before. It is a great tool to highlight the requirement. It's worth to Spend some time to implement it.

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (Alastair-L) [inline packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:140]:
```suggestion  
  async getParametersFromInput(countryCode, entityCode, countryId, entityIds, surveyResponseId){
    if (countryCode) return this.getVariablesByCountryCode(countryCode);
    if (entityCode) return this.getVariablesByEntityCode(entityCode);
    // ...
    throw new ValidationError(
      'Please specify either surveyResponseId, countryId, countryCode, facilityCode or entityIds',
    );
  }
```
Could we shorten this a bit?

File: packages/utils/src/addExportDateAndOriginInExcelExportData.js
Comment (Alastair-L) [inline packages/utils/src/addExportDateAndOriginInExcelExportData.js:16]: Usually I think it's a bit strange to modify the function parameter... could we avoid that here? One option for that:
```suggestion
  return [
    ...exportData,
    [],
    [],
    [`Data exported from Tupaia.org on ${moment().format('Do MMM YYYY')}`]
   ];
```

**Review by Alastair-L — CHANGES_REQUESTED** [review]
Just a few more quick suggestions :). Looking good!

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (billli0) [reply packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:140]: Nice! Would love to change to your approach.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (billli0) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js:99]: Thank you Alastair for the lesson! This is very useful!

File: packages/utils/src/addExportDateAndOriginInExcelExportData.js
Comment (billli0) [reply packages/utils/src/addExportDateAndOriginInExcelExportData.js:16]: Good idea!

**Review by Alastair-L — COMMENTED** [review]
Love it! Can I put it in the reviews channel again for a second look as this has grown in scope a bit and I would prefer to get a second look from someone a bit more familiar with this code?

**Comment by billli0** [comment]
Sure! The logic is a bit complex, that's a good idea to double check.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (tcaiger) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:12]: Please remove ValidationError as it's not used (it was already here before the PR but might as well clean it up)

File: packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js
Comment (tcaiger) [inline packages/meditrak-server/src/routes/utilities/SurveyResponseVariablesExtractor.js:108]: Please use a new variable (const) here rather than reassigning to countryId as it's best practise not to mutate function parameters.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (tcaiger) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:91]: Nice work! I think this is a good abstraction that makes it much easier to read the code.

File: packages/meditrak-server/src/routes/exportSurveyResponses.js
Comment (tcaiger) [inline packages/meditrak-server/src/routes/exportSurveyResponses.js:250]:
```suggestion
      exportData = addExportedDateAndOriginAtTheSheetBottom(exportData);
exportData = [
        ...exportData,
        [],
        [],
        [`Data exported from Tupaia.org on ${moment().format('Do MMM YYYY')}`],
      ];
```

Comment (tcaiger) [reply packages/meditrak-server/src/routes/exportSurveyResponses.js:250]: This is an optional suggestion: I would just add the date and origin inline here and remove the util file. As I see it the advantages to this are:
- the 'addExportedDateAndOriginAtTheSheetBottom' function is very specific and is not really a utility that will be used anywhere else
- it is a simple addition that is easier to read inline
- overall this export code is 'procedural' so it is in general one off code that is written inline
- less overall code to maintain

**Review by tcaiger — APPROVED** [review]
Nice work - looks good to me. I just left a couple of tiny suggestions but nothing that prevents me from approving the PR.

**Review by Alastair-L — APPROVED** [review]
Approving now that we've had a second review.

**Pull request merged**
